### PR TITLE
177403649 mphase gate

### PR DESCRIPTION
**Pull request by mstechly**

### CODE REVIEW

File: src/python/zquantum/core/wip/circuits/_wavefunction_operations.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+from numbers import Number
+
+from ._gates import _sub_symbols, Parameter
+from typing import Tuple, Sequence
+import numpy as np
+
+
+@dataclass(frozen=True)
+class MultiPhaseOperation:
+    """ TODO """
```
Comment (alexjuda) [inline src/python/zquantum/core/wip/circuits/_wavefunction_operations.py:11]: Should this TODO be _todone_?

File: src/python/zquantum/core/wip/circuits/_wavefunction_operations.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    params: Tuple[Parameter, ...]
+
+    def __post_init__(self):
+        if any(
+            isinstance(param, Number) and param.imag != 0
```
Comment (alexjuda) [inline src/python/zquantum/core/wip/circuits/_wavefunction_operations.py:17]: Interesting, I didn't know all Python numbers have this `.imag` property, nice!

File: tests/zquantum/core/wip/circuits/_wavefunction_operations_test.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        [
+            np.array([1, 0, 0, 0]),
+            np.array([1, 1, 0, 0]) / np.sqrt(2),
+            np.array([1, 0, 0, 1]) / np.sqrt(2),
+            np.array([1j, 1j, 1j, 1j]) * 0.5,
+            np.array([0, 1]),
+            np.array([1, 0, 1, 0, 0, 1, 1, 0]) / 2,
+        ],
+    )
+    def test_multi_phase_operation_with_all_parameters_set_to_0_behaves_like_identity(
```
Comment (alexjuda) [inline tests/zquantum/core/wip/circuits/_wavefunction_operations_test.py:19]: `test_multi_phase_operation` in these methods contains the same information as the class name (`TestMultiPhaseOperation`). WDYT about omitting it from the test case name, like:
`test_with_all_parameters_set_to_0_behaves_like_identity`?

File: src/python/zquantum/core/wip/circuits/_wavefunction_operations.py (hunk 1 above)
Comment (alexjuda) [reply src/python/zquantum/core/wip/circuits/_wavefunction_operations.py:11]: Just some ideas: it'd be cool to some reference the origins of this gate / any external resource a future developer might use to validate that the implementation works as expected.

**Review by alexjuda — CHANGES_REQUESTED** [review]

File: src/python/zquantum/core/wip/circuits/_wavefunction_operations.py (hunk 1 above)
Comment (dexter2206) [reply src/python/zquantum/core/wip/circuits/_wavefunction_operations.py]: While we are not aware of any literature covering this operation, I added a description to the docstring in 4aabd68
@alexjuda and @mstechly please see if it is correct and sufficient.

File: src/python/zquantum/core/wip/circuits/_wavefunction_operations.py (hunk 2 above)
Comment (dexter2206) [reply src/python/zquantum/core/wip/circuits/_wavefunction_operations.py:21]: 🐍

File: tests/zquantum/core/wip/circuits/_wavefunction_operations_test.py (hunk 3 above)
Comment (dexter2206) [reply tests/zquantum/core/wip/circuits/_wavefunction_operations_test.py]: Addressed in ab84574

File: src/python/zquantum/core/wip/circuits/_wavefunction_operations.py (hunk 1 above)
Comment (mstechly) [reply src/python/zquantum/core/wip/circuits/_wavefunction_operations.py]: Looks good to me!

**Review by alexjuda — APPROVED** [review]

**Pull request merged**